server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,650 @@
+    ARCHIVE, INCOMING, RESULTS and USERS hierarchy
+    """
+    pbdir = name
+    pbdir_p = os.path.realpath(pbdir)
+
+    if not pbdir_p:
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    if not os.path.isdir(pbdir_p):
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    return 0
+
+
+def main():
+    cfg_name = os.environ.get("_PBENCH_SERVER_CONFIG")
+    if not cfg_name:
+        print(
+            "{}: ERROR: No config file specified; set CONFIG env variable or"
+            " use --config <file> on the command line".format(_NAME_),
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = PbenchServerConfig(cfg_name)
+    except BadConfig as e:
+        print("{}: {} (config file {})".format(_NAME_, e, cfg_name), file=sys.stderr)
+        return 1
+
+    if check_func(config.ARCHIVE, "ARCHIVE") > 0:
+        return 1
+
+    if check_func(config.INCOMING, "INCOMING") > 0:
+        return 1
+
+    if check_func(config.RESULTS, "RESULTS") > 0:
+        return 1
+
+    if check_func(config.USERS, "USERS") > 0:
+        return 1
+
+    logger = get_pbench_logger(_NAME_, config)
+
+    ret = 0
+
+    try:
+        os.mkdir(os.path.join(config.LOGSDIR, "pbench-audit-server"))
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        print("os.mkdir: Unable to create destination directory")
+        return 1
+    logfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.log")
+    # TEMPORARY addition of error file for the sake of test cases
+    errorfile = Path(config.LOGSDIR, "pbench-audit-server", "pbench-audit-server.error")
+    with errorfile.open(mode="w") as f:
+        pass
+    # END
+
+    logger.info("start-{}", config.TS)
+
+    with logfile.open(mode="w") as f:
+
+        ahier = ArchiveHierarchy("archive", config.ARCHIVE, config)
+        verify_archive(ahier)
+        cnt = 0
+        if ahier.check_controller():
+            f.write(f"\nstart-{config.TS[4:]}: archive hierarchy: {config.ARCHIVE}\n")
+            cnt = ahier.dump(f)
+            ahier.header(f, "end")
+        if cnt > 0:
+            ret += 1
+
+        ihier = IncomingHierarchy("incoming", config.INCOMING, config)
+        cihier = ControllerHierarchy("incoming", config.INCOMING, config)
+        if verify_controllers(ihier, cihier) > 0:
+            return 1
+        cnt = check_and_dump(f, cihier, ihier)
+        if cnt > 0:
+            ret += 1
+
+        rhier = ResultsHierarchy("results", config.RESULTS, config)
+        crhier = ControllerHierarchy("results", config.RESULTS, config)
+        if verify_controllers(rhier, crhier) > 0:
+            return 1
+        cnt = check_and_dump(f, crhier, rhier)
+        if cnt > 0:
+            ret += 1
+
+        """
+        ruhier being a ResultsHierarchy object needs to be created
```

Comment:
I think what we really need to document here is that the hierarchy objects are mutated by the validation process (for example, by appending to list properties); so we either need to be able to reliably restore them to pristine state, or throw them away and recreate them. Since most of the cost is in the mutation rather than in creating the objects, it's cheaper to recreate them.